library/src/main/kotlin/com/livefront/gsonkotlinadapter/KoltinReflectiveTypeAdapterFactory.kt

```diff
@@ -40,61 +40,56 @@ class KotlinReflectiveTypeAdapterFactory private constructor(
+        val primaryConstructor: KFunction<T> = kotlinRawType
+            .primaryConstructor
+            ?.apply { isAccessible = true }
+            ?: return null
+        val declaringClass: Class<T> = primaryConstructor.javaConstructor!!.declaringClass!!
+        val innerAdapters: MutableMap<KParameter, TypeAdapter<*>> = mutableMapOf()
+        val constructorParameterDefaultsMap: MutableMap<KParameter, Any?> = mutableMapOf()
+        val invalidReadParameters: MutableSet<KParameter> = mutableSetOf()
+        val constructorMap: MutableMap<String, KParameter> = mutableMapOf()
 
-        private val invalidReadParameters: Set<KParameter> = constructorParameterNameMap
-            .filter { (parameter, names) ->
-                if (enableDefaultPrimitiveValues && parameter.isPrimitive) {
-                    false
+        primaryConstructor.parameters.forEach {
+            val names: List<String> = it.getSerializedNames(declaringClass)
+            if (names.isNotEmpty()) {
+                innerAdapters[it] = gson.getAdapter(type.resolveParameterType(it))
+            }
+            if (!it.isOptional) {
+                constructorParameterDefaultsMap[it] = if (enableDefaultPrimitiveValues) {
+                    it.defaultValue
                 } else {
-                    names.isEmpty() && !parameter.isOptional
+                    null
                 }
-            }
-            .keys
 
-        private val constructorMap: Map<String, KParameter> = constructorParameterNameMap
-            .entries
-            .flatMap { (parameter, names) -> names.map { it to parameter } }
-            .toMap()
+                if (names.isEmpty() && (!it.isPrimitive || !enableDefaultPrimitiveValues)) {
+                    invalidReadParameters += it
+                }
+            }
+            names.forEach { name -> constructorMap[name] = it }
+        }
 
-        private val delegateAdapter: TypeAdapter<T> = gson.getDelegateAdapter(factory, type)
-        private val innerAdapters: Map<KParameter, TypeAdapter<*>> = constructorParameterNameMap
-            .filterNot { (_, names) -> names.isEmpty() }
-            .keys
-            .associateWith { gson.getAdapter(type.resolveParameterType(it)) }
+        return Adapter(
+            gson.getDelegateAdapter(this, type),
+            innerAdapters,
+            kotlinRawType,
+            primaryConstructor,
+            constructorParameterDefaultsMap,
+            invalidReadParameters,
+            constructorMap
+        )
+    }
 
+    internal class Adapter<T : Any>(
+        private val delegateAdapter: TypeAdapter<T>,
+        private val innerAdapters: Map<KParameter, TypeAdapter<*>>,
+        private val kClass: KClass<T>,
+        private val primaryConstructor: KFunction<T>,
+        private val constructorParameterDefaultsMap: Map<KParameter, Any?>,
+        private val invalidReadParameters: Set<KParameter>,
+        private val constructorMap: Map<String, KParameter>
```

Comment:
👍 

So is _all reflection_ gone in what's below now? If so that would really help to make this reusable for a code-gen version of the factory. If not you might want to look into doing so at some point if possible.

Reply:
The `CallBy` is still there and so is the `isSealed`, `isAbstract` and `simpleName`.

I could pass in the `isSealed`, `isAbstract` and `simpleName` as parameters, but the `CallBy` has to stay.

Reply:
Well technically the `CallBy` could be wrapped by a lambda....the reflective case would use the actual `CallBy` while the code-gen case would use the generated constructor. But you can pull all those out later when you get to the code gen stuff, since you'll need to change a bunch of stuff anyway. So this is fine for now then.